src/dyn_model.jl

```diff
@@ -0,0 +1,307 @@
+# Code for building and running a power management problem
+
+
+# ===
+# POWER MANAGEMENT PROBLEM
+# ===
+
+TAU = 1e-5
+INIT_COND = 0
+
+mutable struct PowerManagementProblem
+    problem::Problem
+    p
+    g
+    s
+    params
+end
+
+"""
+    PowerManagementProblem(f, d, pmax, gmax, A, ds; τ=1e-5)
+
+Set up a power management problem with generation costs `f`, demand `d`,
+maximum power flow `pmax`, maximum generation `gmax`, and network
+incidence matrix `A`. `ds` is the change in battery state of charge. It should
+be set to 0 if no battery is in the network. 
+
+The parameter `τ` is a regularization weight used to make the problem
+strongly convex by adding τ ∑ᵢ pᵢ² to the objective. Parameters fq and fl 
+are the Quadratic and Linear terms in the generator costs. P is the maximum battery charge and 
+discharge power (i.e. the maximum amount of energy that can be exchanged with a nodal battery
+during a given timestep).
+
+TODO: check that we're not messing up the ordering!
+"""
+function PowerManagementProblem(fq, fl, d, pmax, gmax, A, ds; τ=TAU)
+    n, m = size(A)
+    g = Variable(n)
+    p = Variable(m)
+
+    # add variable-specific constraints
+    add_constraints(p, [
+        -p <= pmax, 
+        p <= pmax, 
+    ])
+    add_constraints!(g, [
+        -g <= 0, 
+        g <= gmax
+    ]
+    )
+
+    # build problem
+    problem = minimize(
+        (1/2)*quadform(g, Diagonal(fq))
+        + fl'g
+        + (τ/2)*sumsquares(p)
+    )
+    # add network constraint
+    add_constraint!(problem, 0 == A*p - g + d + ds)
+
+    params = (fq=fq, fl=fl, d=d, pmax=pmax, gmax=gmax, A=A, τ=τ)
+
+    # individual PMPs do not have s assigned
+    return PowerManagementProblem(problem, p, g, nothing, params)
+end
+
+# ===
+# Dynamic PowerManagementProblem
+# ===
+"""
+Wrapper and PowerManagementProblem to build a sequence of these. 
+
+The arguments to DynPowerManagementProblem are the same as for PowerManagementProblem
+except they are arrays, with dimension equal to the time horizon (i.e. number of timesteps) T. 
+
+P is the maximum charge/discharge power. 
+C is the maximum SOC of the batteries.
+
+TODO: parameterize the management of initial and terminal constraints
+"""
+function DynPowerManagementProblem(
```

Comment:
Being picky here. In Julia, the convention has been to (almost always) use the full name. I don't think `DynamicPowerManagementProblem` is too much worse. Feel free to ignore this suggestion though.

```suggestion
function DynamicPowerManagementProblem(
```